### PR TITLE
Linting of Python Code + Minor Updates

### PR DESCRIPTION
**Pull request by ghost**
# Summary

This PR takes a first pass at linting and formatting the Python code in the repo 

# Changes
- A number of the ignored `ruff` errors were removed - followed by automated fixes by `ruff` to files with fixable errors/warnings
- Manual Python formatting throughout the repo
- The `test/` directory `README` was updated to match the syntax and format of the top-level `README`
- A couple of minor edits were made to the top-level `README`
- The CI was updated to run more often (e.g., on open, close, and new pushes to a PR) - it appears as if we don't have any usage constraints with the standard hosted GitHub 'runners'
- CI build and lint jobs have been split into two separate jobs.

### CODE REVIEW

File: src/impa/environmentModule.py [hunk 1]
```diff
@@ -4,41 +4,41 @@
 # (See accompanying LICENSE file or at
 #  https://opensource.org/licenses/MIT)
 
+from python_tsp.exact import solve_tsp_dynamic_programming, solve_tsp_brute_force
```
Comment (RustomAlexios) [inline src/impa/environmentModule.py:7]: solve_tsp_dynamic_programming is not used in the code

File: test/python_ksat/src/environmentModule.py [hunk 2]
```diff
@@ -4,73 +4,62 @@
...
-    inf,
-)
 
 # import bitstring
-import argparse
-from collections import (
-    defaultdict,
-)
-
+import multiprocessing as mp
```
Comment (RustomAlexios) [inline test/python_ksat/src/environmentModule.py:9]: not used

File: test/python_ksat/src/environmentModule.py [hunk 3]
```diff
@@ -4,73 +4,62 @@
...
-    tqdm,
-)
+import pdb
+import re
+import os
 
+from multiprocessing.sharedctypes import (
+    Value,
+    Array,
+)
```
Comment (RustomAlexios) [inline test/python_ksat/src/environmentModule.py:26]: not used

File: test/python_ksat/src/environmentModule.py [hunk 4]
```diff
@@ -4,73 +4,62 @@
...
 from multiprocessing import (
     Process,
     Manager,
 )
```
Comment (RustomAlexios) [inline test/python_ksat/src/environmentModule.py:30]: not used

File: test/python_ksat/src/environmentModule.py [hunk 5]
```diff
@@ -4,73 +4,62 @@
...
-
-from multiprocessing.sharedctypes import (
-    Value,
-    Array,
+from collections import (
+    defaultdict,
+)
+from statistics import (
+    mean,
+)
```
Comment (RustomAlexios) [inline test/python_ksat/src/environmentModule.py:36]: not used

File: src/impa/environmentModule.py [hunk 6]
```diff
@@ -4,41 +4,41 @@
...
+from python_tsp.heuristics import solve_tsp_simulated_annealing
+from itertools import combinations, permutations, chain, product
+from ctypes import Structure, c_double
+from collections import defaultdict
+from itertools import islice
+from statistics import mean
```
Comment (RustomAlexios) [inline src/impa/environmentModule.py:13]: not used

File: test/python_kc_mwm/src/environmentModule.py [hunk 7]
```diff
@@ -4,29 +4,30 @@
 # (See accompanying LICENSE file or at
 #  https://opensource.org/licenses/MIT)
 
+from itertools import combinations, permutations
+from ctypes import Structure, c_double
+from statistics import mean
```
Comment (RustomAlexios) [inline test/python_kc_mwm/src/environmentModule.py:9]: not used

File: test/python_tsp/src/environmentModule.py [hunk 8]
```diff
@@ -4,81 +4,62 @@
...
+import math
+import pdb
+import os
+import re
+# import bitstring
 
+from multiprocessing.sharedctypes import (
+    Value,
+    Array,
+)
```
Comment (RustomAlexios) [inline test/python_tsp/src/environmentModule.py:26]: not used

File: test/python_tsp/src/environmentModule.py [hunk 9]
```diff
@@ -4,81 +4,62 @@
...
 from multiprocessing import (
     Process,
     Manager,
 )
```
Comment (RustomAlexios) [inline test/python_tsp/src/environmentModule.py:30]: not used

File: test/python_tsp/src/environmentModule.py [hunk 10]
```diff
@@ -4,81 +4,62 @@
...
-
-from multiprocessing.sharedctypes import (
-    Value,
-    Array,
+from collections import (
+    defaultdict,
+)
+from statistics import (
+    mean,
+)
```
Comment (RustomAlexios) [inline test/python_tsp/src/environmentModule.py:36]: not used

Comment (brendanbcaliola) [reply test/python_tsp/src/environmentModule.py:36]: Are you doing some sore of automated check to see if these are unused imports or are you checking manually?

Comment (RustomAlexios) [reply test/python_tsp/src/environmentModule.py:27]: I recall that some were not used anymore, and if I want to double-check, I do it manually.

File: test/python_tsp/src/environmentModule.py [hunk 11]
```diff
@@ -4,81 +4,53 @@
...
 
-from copy import (
-    deepcopy,
-)
-from cmath import (
-    inf,
-)
-
-# import bitstring
+import multiprocessing as mp
```
Comment (RustomAlexios) [inline test/python_tsp/src/environmentModule.py:7]: not used

File: src/impa/environmentModule.py [hunk 12]
```diff
@@ -4,13 +4,9 @@
 # (See accompanying LICENSE file or at
 #  https://opensource.org/licenses/MIT)
 
-from python_tsp.exact import solve_tsp_dynamic_programming, solve_tsp_brute_force
-from python_tsp.heuristics import solve_tsp_simulated_annealing
 from itertools import combinations, permutations, chain, product
-from ctypes import Structure, c_double
-from collections import defaultdict
-from itertools import islice
```
Comment (RustomAlexios) [inline src/impa/environmentModule.py]: For example, this is used in the code in src/impa/Impa.py

File: src/impa/environmentModule.py (hunk 6 above)
Comment (RustomAlexios) [reply src/impa/environmentModule.py]: We only need to remove "from statistics import mean"

File: src/impa/environmentModule.py [hunk 13]
```diff
@@ -4,41 +4,36 @@
...
 np.set_printoptions(threshold=np.inf)
 np.set_printoptions(suppress=True)
+
 # np_impa_lib = np.float32
 np_impa_lib = np.float64
 zero_value = np_impa_lib(0)
 
 #np.random.seed(17)
-#random.seed(17)
+#random.seed(17)
```
Comment (RustomAlexios) [inline src/impa/environmentModule.py:39]: To resolve the error, we need to add:

from itertools import islice

File: test/python_kc_mwm/src/environmentModule.py [hunk 14]
```diff
@@ -4,29 +4,26 @@
...
 np.set_printoptions(linewidth=np.inf)
 np.set_printoptions(threshold=np.inf)
+
 # np_impa_lib = np.float32
 np_impa_lib = np.float64
-
 zero_value = np_impa_lib(0)
-# np.random.seed(17)
+
+# np.random.seed(17)
```
Comment (RustomAlexios) [inline test/python_kc_mwm/src/environmentModule.py:29]: To resolve the error, we need to add:

from itertools import combinations, permutations

Comment (brendanbcaliola) [reply test/python_kc_mwm/src/environmentModule.py:30]: With regards to your email: 

I've added back this import statement, along with the one in the previous comment. I've tested the Header-only demos, pure python code, python wrapper code, and each of the unit test scripts and everything looks okay on my end. 

Interestingly enough I was not getting errors prior to adding back these imports, so perhaps you want to double check once more on your end that everything looks okay still

Comment (RustomAlexios) [reply test/python_kc_mwm/src/environmentModule.py:30]: Now things are working well on my end. Whenever I make changes to the code locally, I run 
'python3 -m pip install . -v'

**Pull request merged**